sys/src/go/cmd/tmpfs/tmpfs.go

```diff
@@ -0,0 +1,344 @@
+	uname string
+
+	// We can't know how big a serialized dentry is until we serialize it.
+	// At that point it might be too big. We save it here if that happens,
+	// and on the next directory read we start with that.
+	oflow []byte
+
+	// Index of next child to return when reading a directory
+	nextChildIdx int
+}
+
+func newFidEntry(entry tmpfs.Entry, uname string) *FidEntry {
+	return &FidEntry{Entry: entry, uname: uname, oflow: nil, nextChildIdx: -1}
+}
+
+// Rversion initiates the session
+func (fs *fileServer) Rversion(msize protocol.MaxSize, version string) (protocol.MaxSize, string, error) {
+	if version != "9P2000" {
+		return 0, "", fmt.Errorf("%v not supported; only 9P2000", version)
+	}
+	return msize, version, nil
+}
+
+// Rattach attaches a fid to the root for the given user.  aname and afid are not used.
+func (fs *fileServer) Rattach(fid protocol.FID, afid protocol.FID, uname string, aname string) (protocol.QID, error) {
+	if afid != protocol.NOFID {
+		return protocol.QID{}, fmt.Errorf("We don't do auth attach")
+	}
+
+	root := fs.archive.Root()
+	fs.setFile(fid, root, uname)
+
+	return root.Qid(), nil
+}
+
+// Rflush does nothing in tmpfs
+func (fs *fileServer) Rflush(o protocol.Tag) error {
+	return nil
+}
+
+// Rwalk walks the hierarchy from fid, with the walk path determined by paths
+func (fs *fileServer) Rwalk(fid protocol.FID, newfid protocol.FID, paths []string) ([]protocol.QID, error) {
+	// Lookup the parent fid
+	parentEntry, err := fs.getFile(fid)
+	if err != nil {
+		return nil, fmt.Errorf("does not exist")
+	}
+
+	if len(paths) == 0 {
+		// Clone fid - point to same entry
+		fs.setFile(newfid, parentEntry.Entry, parentEntry.uname)
+		return []protocol.QID{}, nil
+	}
+
+	walkQids := make([]protocol.QID, len(paths))
+
+	currEntry := parentEntry.Entry
+	for i, pathcmp := range paths {
+		var ok bool
+		var dir *tmpfs.Directory
+		if dir, ok = currEntry.(*tmpfs.Directory); ok {
+			if pathcmp == ".." {
+				currEntry = dir.Parent()
+			} else {
+				currEntry, ok = dir.ChildByName(pathcmp)
+				if !ok {
+					// From the RFC: If the first element cannot be walked for any
+					// reason, Rerror is returned. Otherwise, the walk will return an
+					// Rwalk message containing nwqid qids corresponding, in order, to
+					// the files that are visited by the nwqid successful elementwise
+					// walks; nwqid is therefore either nwname or the index of the
+					// first elementwise walk that failed. The value of nwqid cannot be
+					// zero unless nwname is zero. Also, nwqid will always be less than
+					// or equal to nwname. Only if it is equal, however, will newfid be
+					// affected, in which case newfid will represent the file reached
+					// by the final elementwise walk requested in the message.
+					//
+					// to sum up: if any walks have succeeded, you return the QIDS for
+					// one more than the last successful walk
+					if i == 0 {
+						return nil, fmt.Errorf("file does not exist")
```

Comment:
consider making consts via errors.New for all these strings, as is common. Try to make sure they match what Linux expects. See if any are in the go syscall package for plan 9 already?